src/tir/transforms/common_subexpr_elim.cc

```diff
@@ -0,0 +1,601 @@
+            computation_and_nb.first, IsEligibleComputation, CanContainEligibleComputations);
+        // The following insertion will maintain `semantic_comp_done_by_expr` sorted (by
+        // decreasing size/complexity), and it will only insert at locations > i as the
+        // direct subexprs are necessarily smaller than the current computation.
+        InsertVectorToSortedSemanticComputations(semantic_comp_done_by_expr, direct_subexprs);
+      }
+    }
+    // Note : we do not remove the current element, as we never look back in the local vector
+  }  // End of for loop
+
+  // Calling the dispatcher to the specific treatments, which will update the context
+  // appropriately before doing the recursive calls on the child nodes
+  result = StmtExprMutator::VisitExpr(result);
+
+  return result;
+}
+
+/*!
+ * \brief The method which overrides the specific treatment for a LetNode
+ */
+PrimExpr CommonSubexpressionEliminator::VisitExpr_(const LetNode* op) {
+  // At this point, we have already done the generic treatment of introducing (via let-in) what
+  // was doable at the toplevel of the given let-in.
+
+  // Save the context at the entry of the function
+  Context context_at_entry = context_;
+
+  // Recurse on the `value` field for potentially rewriting it
+  PrimExpr value_new = VisitExpr(op->value);
+
+  // Augment the context with the association (`var`, `value`) for preparing the next recursion
+  // on the `body`
+  context_.push_back({op->var, MaybeValue(op->value)});
+
+  // Recurse on the `body` (with this extended context)
+  // The recursive call will have potentially done new simplifications, because in this recursive
+  // call `var` will be a part of the context.
+  // (see in VisitExpr() that no introduction were performed when a computation was using an
+  // undefined variable, as that would lead to ill-formed code)
+  PrimExpr body_new = VisitExpr(op->body);
+
+  // Restaure the context to its content at the entrance to not carry out of scope declarations
+  // as the variable introduced by the let-in is not in scope outside of its body
+  context_ = context_at_entry;
+
+  // Rebuild the let-in with a new `value_new` and `body_new` where new simplifications might
+  // have been done.
+
+  // If the `value` and the `body` of the let-in have been rewritten to the same thing
+  if (value_new.same_as(op->value) && body_new.same_as(op->body)) {
+    // then return a reference to the same node
+    return GetRef<PrimExpr>(op);
+  } else {
+    // Otherwise return a let-in built with the new `value_new` and the new `body_new` that
+    // have just been obtained
+    return Let(op->var, value_new, body_new, op->span);
+  }
+}
+
+/*!
+ * \brief The method which overrides the generic dispatcher of StmtExprMutator.
+          Entry point to the common subexpression elimination mutator for statements.
+ * \param stmt The statement to mutate.
+ */
+Stmt CommonSubexpressionEliminator::VisitStmt(const Stmt& stmt) {
```

Comment:
Please remove duplication with `CommonSubexpressionEliminator::VisitExpr` if possible (use template etc)

Reply:
Do we need this specialization for `Stmt` at all? `StmtExprMutator::VisitStmt` visits sub `Expr`s, I think.

Reply:
Yes we do need both the `VisitExpr()` and `VisitStmt()` as we are doing commoning both inside expressions (think about `let x = bigComp+bigComp`) and inside statements (think about `Mem[i] = bigComp+bigComp`).

The basic implementation `StmtExprMutator::VisitStmt` does indeed visit sub `Expr`s. And actually we do rely on `StmtExprMutator::VisitExpr()` and `StmtExprMutator::VisitStmt()` for recursing inside the children, see the end of the function [here](https://github.com/apache/tvm/pull/9482/files#diff-2cfef2190c8093f3b70d5ed20a1e9e75a74cd040a6a0987d5648f511949efce9R307) and [here](https://github.com/apache/tvm/pull/9482/files#diff-2cfef2190c8093f3b70d5ed20a1e9e75a74cd040a6a0987d5648f511949efce9R477).

The algorithm implemented for performing commoning inside expressions and inside statements is of course basically the same (it identifies redundancies, creates variables and then builds new expression/statement using let-in), but many things will be different type-wise when treating a statement versus an expression.
There are many of these subtle differences, but one of them is that for an expression the pass will build a let-in expression and for a statement it will build a let-in statement.

I also thought that it should be possible to factorize them and to replace them by something polymorphic. But when I actually tried a little bit (quite some time ago) I ended up not sure that this is doable easily without putting complexity everywhere.

In particular, a possible approach would be to:
- 1 - I create a main function for the algorithm currently duplicated in `VisitStmt()` and in `VisitExpr()`. Let's call it "`Visit()`" for now. This function takes an `Either<Expr,Stmt>`and returns an `Either<Expr,Stmt>`. Actually, that is called std::variant in C++.
- 2 - In `VisitExpr(const PrimExpr& expr)` : we now build an `std::variant<Expr, Stmt>` from the given `expr`, then call the "generic" `Visit()`, and unpack its result (an Expr) to return it.
- 3 - Same thing for `VisitStmt(const Stmt& stmt)` : we now build an std::variant<Expr, Stmt> from `stmt`, then call the "generic" `Visit()`, and unpack its result (a Stmt) to return it.

In order to achieve 1 (which is the hard part here), we will need to :
- use polymorphic functions for all the functions being used by the current `VisitStmt()` and `VisitExpr()` (i.e. functions with the same name for treating a `PrimExpr `and a `Stmt`). It is already the case for `GetComputationsDoneBy()`, but we need to do the same for `ReplaceSelectedExprInStmt()` and `ReplaceSelectedExprInExpr()` which are currently not polymorphic methods. And the same goes for quite a few other methods.
This part is absolutely not the problem.

- But now, when we try to write the new `Visit()` function, it seems that we are a bit stuck :
We get as input this expr_or_stmt of type std::variant<Expr, Stmt>.
We can of course check what we got :
Expr expr;
Stmt stmt;
if( isExpr(expr_or_stmt) )
  expr = std::get<Expr>(expr_or_stmt);
else
  stmt = std::get<Stmt>(expr_or_stmt);

And then what?
For instance, by what should we replace the beginning of the algorithm, which currently does:
ComputationsDoneBy::GetComputationsDoneBy(
      expr, IsEligibleComputation, CanContainEligibleComputations);
      
Sure, we could check if we're working with an Expr or a Stmt and within a conditional call either 
```
GetComputationsDoneBy(
      **expr**, IsEligibleComputation, CanContainEligibleComputations);
```
or
```
GetComputationsDoneBy(
      **stmt**, IsEligibleComputation, CanContainEligibleComputations);
```
depending on which case we're in. But that replaces a duplication of methods by duplication **within** the method... And as the algorithm was not trivial to follow even with the comments, I though it's probably better to not add complexity in it.

I will try to think about that again soon, but my current way of seeing things is that without adding complexity everywhere, it's not so easy to actually replace these two methods by just a single one doing the job.
In addition, in its current form, it's possible to easily change the algorithm if we want it to do something slightly different for an Expr versus for a Stmt. That would be more complicated if everything goets unified.

Reply:
Thanks for the detailed write-up.

Reply:
My pleasure! Many thanks for your very interesting review! I just have two last thing to do :
- Fix the boolean not needed that you mentionned (last point not resolved).
- Run clang-format again, and final cleanups.